control-plane-operator/hostedclusterconfigoperator/controllers/machine/machine.go

```diff
@@ -0,0 +1,137 @@
+			}
+		}
+	}
+	log.Info("Reconciled Machine")
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) reconcileKubevirtDefaultIngressEndpoints(ctx context.Context, hcp *hyperv1.HostedControlPlane, machine *capiv1.Machine) error {
+	log := ctrl.LoggerFrom(ctx)
+	var namespace string
+	if hcp.Spec.Platform.Kubevirt.Credentials != nil {
+		namespace = hcp.Spec.Platform.Kubevirt.Credentials.InfraNamespace
+	} else {
+		namespace = hcp.Namespace
+	}
+
+	// Manifests for infra/mgmt cluster passthrough service
+	cpService := hcpmanifests.IngressDefaultIngressPassthroughService(namespace)
+
+	cpService.Name = fmt.Sprintf("%s-%s",
+		hcpmanifests.IngressDefaultIngressPassthroughServiceName,
+		hcp.Spec.Platform.Kubevirt.GenerateID)
+
+	if err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(cpService), cpService); err != nil {
+		return fmt.Errorf("failed to get default ingress passthrow Service: %w", err)
+	}
+
+	// If there is a selector endpoints should not be generated
+	if len(cpService.Spec.Selector) > 0 {
+		return nil
+	}
+
+	if len(cpService.Spec.Ports) == 0 {
+		return fmt.Errorf("missing default ingress passthrow Service %s/%s ports", cpService.Namespace, cpService.Name)
+	}
+
+	ipv4EndpointSlice := &discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cpService.Namespace,
+			Name:      cpService.Name + "-" + machine.Name + "-ipv4",
+			Labels: map[string]string{
+				"kubernetes.io/service-name": cpService.Name,
+			},
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: machine.APIVersion,
+				Kind:       machine.Kind,
+				Name:       machine.Name,
+				UID:        machine.UID,
+			}},
+		},
+		AddressType: discoveryv1.AddressTypeIPv4,
+	}
+
+	result, err := r.CreateOrUpdate(ctx, r.kubevirtInfraClient, ipv4EndpointSlice, func() error {
+		err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(ipv4EndpointSlice), ipv4EndpointSlice)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get default ingress passthrow Endpoints: %w", err)
+		}
+		endpoint := discoveryv1.Endpoint{
+			Addresses: []string{},
+		}
+		for _, machineAddress := range machine.Status.Addresses {
+			if machineAddress.Type == capiv1.MachineInternalIP {
+				if machine.Status.GetTypedPhase() == capiv1.MachinePhaseRunning {
+					if netip.MustParseAddr(machineAddress.Address).Is4() {
+						endpoint.Addresses = append(endpoint.Addresses, machineAddress.Address)
+					}
+				}
+			}
+		}
+
+		ports := []discoveryv1.EndpointPort{}
+		for _, port := range cpService.Spec.Ports {
+			ports = append(ports, discoveryv1.EndpointPort{
+				Protocol: &port.Protocol,
+				Port:     pointer.Int32(int32(port.TargetPort.IntValue())),
+			})
+		}
+
+		if len(endpoint.Addresses) == 0 {
+			return fmt.Errorf("missing machine addresses to compose default ingress passthrow endpointslices")
```

Comment:
typo, passthrow/passthrough

Also, I'm not completely sure this is an error condition. There are many reasons why there might not be an address yet and that shouldn't be considered a reconcile error.

For example, if a machine is not yet in a running phase, we don't want to set an address, but that's also not a reconcile error. Also, if the machine is in a failed phase, then we want to remove the address from the endpoint so traffic isn't sent to the failed node. That's also a situation where address len is 0 and it's correct.

Reply:
Removed the error, now when the machine reaches "phase" Running the endpointslices are created.